### PR TITLE
Add support for launching on-demand AWS instances

### LINKED ISSUE #25: Add support for on-demand AWS instance
I have a horrible hacky implementation on #23: https://github.com/znewman01/tsunami/commit/8c91e80513f6c587621d2ac30a88e8ed046e2430

Which I can clean up (at some point)

### PR DESCRIPTION
**Pull request by jonhoo**
I've done a small-scale test of it, and it seems to be working!

Fixes #25.

/cc @akshayknarayan @znewman01

### CODE REVIEW

File: src/providers/aws.rs [hunk 1]
```diff
@@ -109,6 +109,38 @@ use std::time;
 use tracing::instrument;
 use tracing_futures::Instrument;
 
+/// Dictate how a set of instances should be launched.
+#[derive(Debug, Clone)]
+#[allow(missing_copy_implementations)]
```
Comment (akshayknarayan) [inline src/providers/aws.rs:115]: What's wrong with having it be copy?

File: src/providers/aws.rs [hunk 2]
```diff
@@ -827,6 +877,118 @@ impl RegionLauncher {
         Ok(self)
     }
 
+    #[instrument(level = "trace", skip(self))]
+    async fn make_on_demand_requests<M>(&mut self, machines: M) -> Result<(), Report>
```
Comment (akshayknarayan) [inline src/providers/aws.rs]: I wonder if it might be possible to share more of this code with the other implementation, but this seems fine.

Comment (akshayknarayan) [reply src/providers/aws.rs]: Maybe a few utility functions to e.g. do the instance grouping and make placement groups?

**Review by akshayknarayan — COMMENTED** [review]
Seems good! We probably want to also update this line of the module docs: https://github.com/jonhoo/tsunami/pull/45/files#diff-60699702750aef5a8cdc050b4fce64eaL7

File: src/providers/aws.rs (hunk 1 above)
Comment (jonhoo) [reply src/providers/aws.rs:115]: In general, making types `Copy` is a backwards-compatibility risk. It means that if we _later_ add a field that can't be `Copy` (like a `String`), then we'd have to do a breaking release. In general, you should only implement `Copy` for public types if you think it is highly unlikely that they might gain non-`Copy` members in the future. For this, I think there's a decent chance we may add a `String` field later, so no `Copy` :)

File: src/providers/aws.rs (hunk 2 above)
Comment (jonhoo) [reply src/providers/aws.rs]: Oh, yeah, probably. I'll do a pass and see if I can tidy it up a little.

File: src/providers/aws.rs [hunk 3]
```diff
@@ -109,6 +110,38 @@ use std::time;
...
+    },
+    /// Use regular AWS on-demand instances.
+    OnDemand,
+}
+
+impl LaunchMode {
+    /// Launch using AWS [defined duration](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances) spot instances.
+    ///
+    /// The lifetime of such instances must be declared in advance (1-6 hours).
+    /// This method thus clamps `hours` to be between 1 and 6.
```
Comment (znewman01) [inline src/providers/aws.rs:132]: This is definitely personal preference but I'd rather barf on invalid input than silently succeed. Totally fine either way as long as it's documented

Comment (znewman01) [reply src/providers/aws.rs:132]: Waiting on those [range types](https://github.com/rust-lang/rfcs/issues/671) I guess :)

File: src/providers/aws.rs [hunk 4]
```diff
@@ -827,6 +880,134 @@ impl RegionLauncher {
         Ok(self)
     }
 
+    #[instrument(level = "trace", skip(self, mk))]
+    async fn make_placement<R>(
```
Comment (znewman01) [inline src/providers/aws.rs:889]: I know this is private but worth documenting briefly since the point of `mk` is a little nonobvious from its type/name.

**Review by znewman01 — APPROVED** [review]
handful of purely-aesthetic comments, main logic looks good to me

i guess i have a longer-term question about how to test this library but it's out-of-scope for now :)

File: src/providers/aws.rs (hunk 3 above)
Comment (jonhoo) [reply src/providers/aws.rs:132]: Yeah, this one was a tricky decision. The main issue I have here is that panicking would mean that cleanup isn't run, which means that _past_ instances may be left running. We could define an `enum` I suppose, but that seems like overkill. My inclination is to keep it as-is.

**Comment by jonhoo** [comment]
Yeah, I also would love to have this library have decent testing. But how to do that properly, I have no idea...

**Pull request merged**